rules/scripts/bitstreams_workspace.py

```diff
@@ -0,0 +1,301 @@
+        """Get the requested bitstream files.
+
+        Args:
+            key: str; A git hash or the string 'latest'.
+        Returns:
+            dict[str:str]: A dictionary mapping the bitstream types to
+                           their target files.
+        """
+        files = self._GetFromLocal(key)
+        if not files:
+            self._GetFromRemote(key)
+            files = self._GetFromLocal(key)
+        return {os.path.splitext(f)[1][1:]: f for f in files}
+
+    def WriteBuildFile(self, build, key):
+        """Write a BUILD file for the requested bitstream files.
+
+        Args:
+            build: path; Filename of the BUILD file to write.
+            key: str; A git hash or the string 'latest'.
+        """
+        param = self.GetFromCache(key)
+        param['datetime'] = datetime.datetime.now().isoformat()
+        param['key'] = key
+
+        with open(build, 'wt') as f:
+            f.write("""# This file was autogenerated. Do not edit!
+# Built at {datetime}.
+# Configured for bitstream: {key}
+
+package(default_visibility = ["//visibility:public"])
+
+exports_files(glob(["cache/**"]))
+
+filegroup(
+    name = "bitstream_test_rom",
+    srcs = ["{orig}"],
+)
+
+filegroup(
+    name = "bitstream_mask_rom",
+    srcs = ["{splice}"],
+)
+""".format(**param))
+
+        if key == 'latest':
+            key = self.available[key]
+        return key
+
+
+def main(argv):
+    # The user can override some basic behaviors with the BITSTREAM
+    # environment variable.
+    env = os.environ.get('BITSTREAM')
+    if env:
+        argv.extend(env.split(' '))
+    args = parser.parse_args(args=argv[1:])
+    bitstream = args.bitstream
+
+    # We need to know the location of the main git repo, since this script
+    # will have its CWD in a bazel-managed 'external' directory.
+    # If not provided, we assume this script itself is located in the main
+    # git repository.
+    if args.repo:
+        if os.path.isdir(args.repo):
+            repo = args.repo
+        else:
+            repo = os.path.dirname(args.repo)
+    else:
+        repo = os.path.dirname(argv[0])
+
+    cache = BitstreamCache(args.bucket_url, args.cache, args.latest_update, args.offline)
+    cache.InitRepository()
+
+    # Do we need a refresh?
+    need_refresh = (args.refresh or
+                    bitstream != 'latest' or
+                    cache.NeedRefresh(args.refresh_time) and not args.offline)
+    cache.GetBitstreamsAvailable(need_refresh)
+
+    # If commanded to print bitstream availability, do so.
+    if args.list:
+        for k, v in cache.available.items():
+            print('{}: {}'.format(k, v))
+
+    # If we aren't getting the latest bitstream, resolve the hash to the closest
+    # bitstream we can find.
+    if bitstream != 'latest':
+        closest = cache.GetClosest(repo, bitstream)
+        if closest is None:
+            print('Cannot find a bitstream close to {}'.format(bitstream))
+            return 2
```

Comment:
minor nit: Is there a reason for return code of 2 vs 1? Just curious, I figured this would fall under the category of ["all other kind of errors."](https://docs.python.org/2/library/sys.html#sys.exit)

Reply:
Typo.  Fixed.